File: apps/web/lib/types/competition.ts
/**
 * The competition layer is sport-agnostic on purpose.
 *
 * A result is one number (`score_value`) in a unit the club names itself
 * (`scoring_unit`: Punkte, Ringe, Sekunden, Meter …), ranked by
 * `scoring_mode`. Anything sport-specific — rings, targets, calibres — lives
 * in `details` and belongs to the module that understands it, never here.
 */

/** Types a club can pick. `free_training` is machinery and never offered. */
export const COMPETITION_TYPE_KEYS = [
  "competition",
  "league",
  "training",
] as const

export type CompetitionType = (typeof COMPETITION_TYPE_KEYS)[number]

export const SCORING_MODE_KEYS = ["highest_wins", "lowest_wins"] as const

export type Competition = {
  id: string
  name: string
  description: string | null
  competition_type: string
  start_date: string
  end_date: string | null
  /** "highest_wins" for points and rings, "lowest_wins" for times. */
  scoring_mode: string
  /** The club's own word for the number, e.g. "Punkte" or "Sekunden". */
  scoring_unit: string
  disciplines: string[] | null
  created_at: string
  updated_at: string
}

/** One round of a competition — a match day, a leg, a training evening. */
export type CompetitionSession = {
  id: string
  competition_id: string
  name: string | null
  date: string
  location: string | null
  discipline: string | null
  /** Set when the round also sits in the calendar. */
  event_id: string | null
  created_at: string
  updated_at: string
}

/** One participant's result in a round. */
export type CompetitionEntry = {
  id: string
  session_id: string
  member_id: string
  score_value: number
  score_unit: string
  discipline: string | null
  details: Record<string, unknown> | null
  /** "manual" | "scan" — scan means it came from a device, not a keyboard. */
  source: string
  recorded_by: string | null
  recorded_at: string
  notes: string | null
}

/** A row of the ranking, as computed by the backend. */
export type ScoreboardRow = {
  member_id: string
  member_name: string
  total_score: number
  entry_count: number
  average_score: number
  best_score: number
  rank: number
}

/** The scoreboard response carries the competition's scale alongside the rows. */
export type Scoreboard = {
  rows: ScoreboardRow[]
  scoring_mode: string
  scoring_unit: string
}

File: apps/mobile/android/core/model/src/main/kotlin/com/unefy/core/model/Competition.kt
package com.unefy.core.model

/**
 * A competition. Sport-agnostic on purpose: the scoring unit and mode come from
 * the backend, so a shooting club counts rings and another club counts whatever
 * it counts without the app knowing the difference.
 */
data class Competition(
    val id: String,
    val name: String,
    val description: String?,
    val type: String?,
    val startDate: String,
    val endDate: String?,
    val scoringUnit: String,
    val scoringMode: String,
    val disciplines: List<String>,
) {
    /** True when higher scores win, which decides how a ranking reads. */
    val highestWins: Boolean get() = scoringMode == "highest_wins"
}

/**
 * A ranking over all sessions of a competition.
 *
 * [unit] travels with the rows rather than being assumed: "1040" means nothing
 * without knowing it is rings.
 */
data class Scoreboard(
    val unit: String,
    val highestWins: Boolean,
    val rows: List<ScoreboardRow>,
)

data class ScoreboardRow(
    val rank: Int,
    val memberId: String,
    val memberName: String,
    val totalScore: Double,
    val bestScore: Double,
    val averageScore: Double,
    val entryCount: Int,
)

/**
 * One round of a competition — a match day, a leg, a training evening.
 *
 * The unit a series is filed under: results belong to a round, and the round
 * belongs to the competition.
 */
data class CompetitionRound(
    val id: String,
    val competitionId: String,
    val name: String?,
    /** ISO date. */
    val date: String,
    val location: String?,
    val discipline: String?,
    /** Set when the round also sits in the calendar. */
    val eventId: String?,
)
